File: exercism/elixir/nth-prime/nth_prime.exs
defmodule Prime do
  @doc """
  Generates the nth prime.
  """
  @spec nth(non_neg_integer) :: non_neg_integer
  def nth(0), do: raise ArgumentError
  def nth(count) do
    primes
    |> Stream.drop(count-1)
    |> Enum.take(1)
    |> hd
  end

  def primes do
    initial = fn -> Stream.concat([2], Stream.iterate(3, fn a -> a + 2 end)) end
    next = fn acc ->
      prime = Enum.take(acc, 1) |> hd
      {[prime], Stream.reject(acc, fn n -> rem(n, prime) == 0 end)}
    end

    Stream.resource(initial, next, fn _ -> end)
  end
end

File: exercism/scala/nth-prime/src/main/scala/nthPrime.scala
object Prime {
  def prime(limit: Int): Int = {
    val step = 2

    @scala.annotation.tailrec
    def find(count: Int, n: Int, primes: List[Int]): Int =
      if (count == limit) n - step
      else {
        if (primes.forall(p => n % p != 0)) find(count + 1, n + step, n :: primes)
        else find(count, n + step, primes)
      }

    find(0, 3, List.empty[Int])
  }

  def nth(n: Int): Int = n match {
    case 1 => 2
    case _ => prime(n-1)
  }
}
